README.md
# Emubox
Customizations for Grub, EmulationStation, Retroarch etc. to be used with Emubox-CD. Uses scripts from RetroPie.


debian/changelog
emubox (1.0) unstable; urgency=medium

  * Initial Release.

 -- Ashwin Menon <ashwin.menon@gmail.com>  Sat, 24 Jun 2017 11:09:07 +0530


data/usr/share/emubox/scripts/inifuncs.sh
#!/bin/bash

# This file is part of The RetroPie Project
#
# The RetroPie Project is the legal property of its developers, whose names are
# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
#
# See the LICENSE.md file at the top-level directory of this distribution and
# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
#

## @file inifuncs.sh
## @brief RetroPie inifuncs library
## @copyright GPLv3

# @fn fatalError()
# @param message string or array of messages to display
# @brief echos message, and exits immediately.
function fatalError() {
    echo -e "$1"
    exit 1
}

# arg 1: delimiter, arg 2: quote, arg 3: file

## @fn iniConfig()
## @param delim ini file delimiter eg. ' = '
## @param quote ini file quoting character eg. '"'
## @param config ini file to edit
## @brief Configure an ini file for getting/setting values with `iniGet` and `iniSet`
function iniConfig() {
    __ini_cfg_delim="$1"
    __ini_cfg_quote="$2"
    __ini_cfg_file="$3"
}

# arg 1: command, arg 2: key, arg 2: value, arg 3: file (optional - uses file from iniConfig if not used)

# @fn iniProcess()
# @param command `set`, `unset` or `del`
# @param key ini key to operate on
# @param value to set
# @param file optional file to use another file than the one configured with iniConfig
# @brief The main function for setting and deleting from ini files - usually
# not called directly but via iniSet iniUnset and iniDel
function iniProcess() {
    local cmd="$1"
    local key="$2"
    local value="$3"
    local file="$4"
    [[ -z "$file" ]] && file="$__ini_cfg_file"
    local delim="$__ini_cfg_delim"
    local quote="$__ini_cfg_quote"

    [[ -z "$file" ]] && fatalError "No file provided for ini/config change"
    [[ -z "$key" ]] && fatalError "No key provided for ini/config change on $file"

    # we strip the delimiter of spaces, so we can "fussy" match existing entries that have the wrong spacing
    local delim_strip=${delim// /}
    # if the stripped delimiter is empty - such as in the case of a space, just use the delimiter instead
    [[ -z "$delim_strip" ]] && delim_strip="$delim"
    local match_re="^[[:space:]#]*$key[[:space:]]*$delim_strip.*$"

    local match
    if [[ -f "$file" ]]; then
        match=$(egrep -i "$match_re" "$file" | tail -1)
    else
        touch "$file"
    fi

    if [[ "$cmd" == "del" ]]; then
        [[ -n "$match" ]] && sed -i -e "\|$(sedQuote "$match")|d" "$file"
        return 0
    fi

    [[ "$cmd" == "unset" ]] && key="# $key"

    local replace="$key$delim$quote$value$quote"
    if [[ -z "$match" ]]; then
        # make sure there is a newline then add the key-value pair
        sed -i '$a\' "$file"
        echo "$replace" >> "$file"
    else
        # replace existing key-value pair
        sed -i -e "s|$(sedQuote "$match")|$(sedQuote "$replace")|g" "$file"
    fi

    [[ "$file" =~ retroarch\.cfg$ ]] && retroarchIncludeToEnd "$file"
}

## @fn iniUnset()
## @param key ini key to operate on
## @param value to Unset (key will be commented out, but the value can be changed also)
## @param file optional file to use another file than the one configured with iniConfig
## @brief Unset (comment out) a key / value pair in an ini file.
## @details The key does not have to exist - if it doesn't exist a new line will
## be added - eg. `# key = "value"`
##
## This function is useful for creating example configuration entries for users
## to manually enable later or if a configuration is to be disabled but left
## as an example.
function iniUnset() {
    iniProcess "unset" "$1" "$2" "$3"
}

## @fn iniSet()
## @param key ini key to operate on
## @param value to set
## @param file optional file to use another file than the one configured with iniConfig
## @brief Set a key / value pair in an ini file.
## @details If the key already exists the existing line will be changed. If not
## a new line will be created.
function iniSet() {
    iniProcess "set" "$1" "$2" "$3"
}

## @fn iniDel()
## @param key ini key to operate on
## @param file optional file to use another file than the one configured with iniConfig
## @brief Delete a key / value pair in an ini file.
function iniDel() {
    iniProcess "del" "$1" "" "$2"
}

## @fn iniGet()
## @param key ini key to get the value of
## @param file optional file to use another file than the one configured with iniConfig
## @brief Get the value of a key from an ini file.
## @details The value of the key will end up in the global ini_value variable.
function iniGet() {
    local key="$1"
    local file="$2"
    [[ -z "$file" ]] && file="$__ini_cfg_file"
    if [[ ! -f "$file" ]]; then
        ini_value=""
        return 1
    fi

    local delim="$__ini_cfg_delim"
    local quote="$__ini_cfg_quote"
    # we strip the delimiter of spaces, so we can "fussy" match existing entries that have the wrong spacing
    local delim_strip=${delim// /}
    # if the stripped delimiter is empty - such as in the case of a space, just use the delimiter instead
    [[ -z "$delim_strip" ]] && delim_strip="$delim"

    # create a regexp to match the value based on whether we are looking for quotes or not
    local value_m
    if [[ -n "$quote" ]]; then
        value_m="$quote*\([^$quote|\r]*\)$quote*"
    else
        value_m="\([^\r]*\)"
    fi

    ini_value="$(sed -n "s/^[ |\t]*$key[ |\t]*$delim_strip[ |\t]*$value_m.*/\1/p" "$file" | tail -1)"
}

# @fn retroarchIncludeToEnd()
# @param file config file to process
# @brief Makes sure a `retroarch.cfg` file has the `#include` line at the end.
# @details Used in runcommand.sh and iniProcess to ensure the #include for the
# main retroarch.cfg is always at the end of a system `retroarch.cfg`. This
# is because when processing its config RetroArch will take the first value it
# finds, so any overrides need to be above the `#include` line where the global
# retroarch.cfg is included.
function retroarchIncludeToEnd() {
    local config="$1"

    [[ ! -f "$config" ]] && return

    local re="^#include.*retroarch\.cfg"

    # extract the include line (unless it is the last line in the file)
    # (remove blank lines, the last line and search for an include line in remaining lines)
    local include=$(sed '/^$/d;$d' "$config" | grep "$re")

    # if matched remove it and re-add it at the end
    if [[ -n "$include" ]]; then
        sed -i "/$re/d" "$config"
        # add newline if missing and the #include line
        sed -i '$a\' "$config"
        echo "$include" >>"$config"
    fi
}

# escape backslashes and pipes for sed
function sedQuote() {
    local string="$1"
    string="${string//\\/\\\\}"
    string="${string//|/\\|}"
    echo "$string"
}


data/usr/share/emubox/emulationstation/scripts/inputconfiguration.sh
#!/usr/bin/env bash

# This file is part of The RetroPie Project
#
# The RetroPie Project is the legal property of its developers, whose names are
# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
#
# See the LICENSE.md file at the top-level directory of this distribution and
# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
#

## @file scriptmodules/supplementary/emulationstation/inputconfiguration.sh
## @brief input configuration script
## @copyright GPLv3
## @details
## @par global variables
##
## There are 3 global variables which are set to the current device being processed
##
## `DEVICE_TYPE` = device type is currently either joystick or keyboard
## `DEVICE_NAME` = name of the device
## `DEVICE_GUID` = SDL2 joystick GUID of the device (-1 for keyboard)
##
## @par Interface functions
##
## each input configuration module can have an optional function
## `check_<filename without extension>`. If this function returns 1, the module is
## skipped. This can be used to skip input configurations in some cases - eg
## to skip configuration when an existing config file is not installed.
##
## There are 3 main interface functions for each of the input types (joystick/keyboard)
##
## function `onstart_<filename without extension>_<inputtype>()`
## is run at the start of the input configuration
##
## function `onend_<filename without extension>_<inputtype>()`
## is run at the end of the input configuration
##
## Returns:
##   None
##
## function map_<filename without extension>_<inputtype>()
## is run for each of the inputs - with the following arguments
##
## Arguments:
## * $1 - input name is one of the following
##  * up, down, left, right
##  * a, b, x, y
##  * leftshoulder, rightshoulder, lefttrigger, righttrigger
##  * leftthumb. rightthumb
##  * start, select
##  * leftanalogup, leftanalogdown, leftanalogleft, leftanalogright
##  * rightanalogup, rightanalogdown, rightanalogleft, rightanalogright
## * $2 - input type for joysticks are button, axis, hat or key for keyboard.
## * $3 - button id of the input for a joystick, or SDL2 keycode for a keyboard.
## * $4 - value of the joystick input or 1 for keyboard.
##
## Returns:
##   None

function inputconfiguration() {

    local es_conf="/home/emubox/.emulationstation/es_temporaryinput.cfg"
    declare -A mapping

    # check if we have the temporary input file
    [[ ! -f "$es_conf" ]] && return

    local line
    while read line; do
        if [[ -n "$line" ]]; then
            local input=($line)
            mapping["${input[0]}"]=${input[@]:1}
        fi
    done < <(xmlstarlet sel --text -t -m "/inputList/inputConfig/input"  -v "concat(@name,' ',@type,' ',@id,' ',@value)" -n "$es_conf")

    local inputscriptdir=$(dirname "$0")
    local inputscriptdir=$(cd "$inputscriptdir" && pwd)

    DEVICE_TYPE=$(xmlstarlet sel --text -t -v "/inputList/inputConfig/@type" "$es_conf")
    DEVICE_NAME=$(xmlstarlet sel --text -t -v "/inputList/inputConfig/@deviceName" "$es_conf")
    DEVICE_GUID=$(xmlstarlet sel --text -t -v "/inputList/inputConfig/@deviceGUID" "$es_conf")

    echo "Input type is '$DEVICE_TYPE'."

    local module
    # call all configuration modules with the
    for module in $(find "$inputscriptdir/configscripts/" -maxdepth 1 -name "*.sh" | sort); do

        source "$module"  # register functions from emulatorconfigs folder
        local module_id=${module##*/}
        local module_id=${module_id%.sh}

        local funcname="check_${module_id}"
        # call check_module to check if we should run it
        if fn_exists "$funcname"; then
            "$funcname" || continue
        fi

        echo "Configuring '$module_id'"

        # at the start, the onstart_module function is called.
        funcname="onstart_${module_id}_${DEVICE_TYPE}"
        fn_exists "$funcname" && "$funcname"

        local input_name
        # loop through all buttons and use corresponding config function if it exists
        for input_name in "${!mapping[@]}"; do
            funcname="map_${module_id}_${DEVICE_TYPE}"

            if fn_exists "$funcname"; then
                local params=(${mapping[$input_name]})
                local input_type=${params[0]}
                local input_id=${params[1]}
                local input_value=${params[2]}

                "$funcname" "$input_name" "$input_type" "$input_id" "$input_value"
            fi
        done

        # at the end, the onend_module function is called
        funcname="onend_${module_id}_${DEVICE_TYPE}"
        fn_exists "$funcname" && "$funcname"

    done

}

function fn_exists() {
    declare -f "$1" > /dev/null
    return $?
}

function sdl1_map() {
    # check if sdl1_map already exists
    [[ -v sdl1_map[@] ]] && return

    # SDL codes from https://wiki.libsdl.org/SDLKeycodeLookup mapped to /usr/include/SDL/SDL_keysym.h
    declare -Ag sdl1_map
    local i
    for i in {0..127}; do
        sdl1_map["$i"]="$i"
    done

    sdl1_map["1073741881"]="301" # SDLK SDLK_CAPSLOCK
    sdl1_map["1073741882"]="282" # SDLK SDLK_F1
    sdl1_map["1073741883"]="283" # SDLK SDLK_F2
    sdl1_map["1073741884"]="284" # SDLK SDLK_F3
    sdl1_map["1073741885"]="285" # SDLK SDLK_F4
    sdl1_map["1073741886"]="286" # SDLK SDLK_F5
    sdl1_map["1073741887"]="287" # SDLK SDLK_F6
    sdl1_map["1073741888"]="288" # SDLK SDLK_F7
    sdl1_map["1073741889"]="289" # SDLK SDLK_F8
    sdl1_map["1073741890"]="290" # SDLK SDLK_F9
    sdl1_map["1073741891"]="291" # SDLK SDLK_F10
    sdl1_map["1073741892"]="292" # SDLK SDLK_F11
    sdl1_map["1073741893"]="293" # SDLK SDLK_F12
    sdl1_map["1073741894"]="316" # SDLK SDLK_PRINTSCREEN
    sdl1_map["1073741895"]="302" # SDLK SDLK_SCROLLLOCK
    sdl1_map["1073741896"]="19"  # SDLK SDLK_PAUSE
    sdl1_map["1073741897"]="277" # SDLK SDLK_INSERT
    sdl1_map["1073741898"]="278" # SDLK SDLK_HOME
    sdl1_map["1073741899"]="280" # SDLK SDLK_PAGEUP
    sdl1_map["1073741901"]="279" # SDLK SDLK_END
    sdl1_map["1073741902"]="281" # SDLK SDLK_PAGEDOWN
    sdl1_map["1073741903"]="275" # SDLK SDLK_RIGHT
    sdl1_map["1073741904"]="276" # SDLK SDLK_LEFT
    sdl1_map["1073741905"]="274" # SDLK SDLK_DOWN
    sdl1_map["1073741906"]="273" # SDLK SDLK_UP
    sdl1_map["1073741908"]="267" # SDLK SDLK_KP_DIVIDE
    sdl1_map["1073741909"]="268" # SDLK SDLK_KP_MULTIPLY
    sdl1_map["1073741910"]="269" # SDLK SDLK_KP_MINUS
    sdl1_map["1073741911"]="270" # SDLK SDLK_KP_PLUS
    sdl1_map["1073741912"]="271" # SDLK SDLK_KP_ENTER
    sdl1_map["1073741913"]="257" # SDLK SDLK_KP_1
    sdl1_map["1073741914"]="258" # SDLK SDLK_KP_2
    sdl1_map["1073741915"]="259" # SDLK SDLK_KP_3
    sdl1_map["1073741916"]="260" # SDLK SDLK_KP_4
    sdl1_map["1073741917"]="261" # SDLK SDLK_KP_5
    sdl1_map["1073741918"]="262" # SDLK SDLK_KP_6
    sdl1_map["1073741919"]="263" # SDLK SDLK_KP_7
    sdl1_map["1073741920"]="264" # SDLK SDLK_KP_8
    sdl1_map["1073741921"]="265" # SDLK SDLK_KP_9
    sdl1_map["1073741922"]="256" # SDLK SDLK_KP_0
    sdl1_map["1073741923"]="266" # SDLK SDLK_KP_PERIOD
    sdl1_map["1073741926"]="320" # SDLK SDLK_POWER
    sdl1_map["1073741927"]="272" # SDLK SDLK_KP_EQUALS
    sdl1_map["1073741928"]="294" # SDLK SDLK_F13
    sdl1_map["1073741929"]="295" # SDLK SDLK_F14
    sdl1_map["1073741930"]="296" # SDLK SDLK_F15
    sdl1_map["1073741941"]="315" # SDLK SDLK_HELP
    sdl1_map["1073741942"]="319" # SDLK SDLK_MENU
    sdl1_map["1073741946"]="322" # SDLK SDLK_UNDO
    sdl1_map["1073741978"]="317" # SDLK SDLK_SYSREQ
    sdl1_map["1073742048"]="306" # SDLK SDLK_LCTRL
    sdl1_map["1073742049"]="304" # SDLK SDLK_LSHIFT
    sdl1_map["1073742050"]="308" # SDLK SDLK_LALT
    sdl1_map["1073742051"]="311" # SDLK SDLK_LGUI
    sdl1_map["1073742052"]="305" # SDLK SDLK_RCTRL
    sdl1_map["1073742053"]="303" # SDLK SDLK_RSHIFT
    sdl1_map["1073742054"]="307" # SDLK SDLK_RALT
    sdl1_map["1073742055"]="312" # SDLK SDLK_RGUI
    sdl1_map["1073742081"]="313" # SDLK SDLK_MODE
}

###### main ######

rootdir="/usr/share/emubox"
source "$rootdir/scripts/inifuncs.sh"
inputconfiguration


data/usr/share/emubox/emulationstation/scripts/configscripts/emulationstation.sh
#!/usr/bin/env bash

# This file is part of The RetroPie Project
#
# The RetroPie Project is the legal property of its developers, whose names are
# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
#
# See the LICENSE.md file at the top-level directory of this distribution and
# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
#

function onstart_emulationstation_joystick() {
    local es_conf="/home/emubox/.emulationstation/es_input.cfg"

    mkdir -p "/home/emubox/.emulationstation"

    if [[ ! -f "$es_conf" ]]; then
        echo "<inputList />" >"$es_conf"
    else
        cp "$es_conf" "$es_conf.bak"
    fi

    # look for existing inputConfig in config by GUID
    if [[ $(xmlstarlet sel -t -v "count(/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"])" "$es_conf") -eq 0 ]]; then
        # if not found by GUID, look for inputConfig with deviceName only
        if [[ $(xmlstarlet sel -t -v "count(/inputList/inputConfig[@deviceName=\"$DEVICE_NAME\"][not(@deviceGUID)])" "$es_conf") -eq 0 ]]; then
            # insert new inputConfig entry
            xmlstarlet ed -L -s "/inputList" -t elem -n newInputConfig -v "" \
                -i //newInputConfig -t attr -n "type" -v "$DEVICE_TYPE" \
                -i //newInputConfig -t attr -n "deviceName" -v "$DEVICE_NAME" \
                -i //newInputConfig -t attr -n "deviceGUID" -v "$DEVICE_GUID" \
                -r //newInputConfig -v inputConfig \
                "$es_conf"
        else
            # add deviceGUID to inputConfig
            xmlstarlet ed -L \
                -i "/inputList/inputConfig[@deviceName=\"$DEVICE_NAME\"]" -t attr -n "deviceGUID" -v "$DEVICE_GUID" \
                "$es_conf"
        fi
    fi
}

function map_emulationstation_joystick() {
    local input_name="$1"
    local input_type="$2"
    local input_id="$3"
    local input_value="$4"

    local key
    case "$input_name" in
        leftbottom|leftshoulder)
            key="pageup"
            ;;
        rightbottom|rightshoulder)
            key="pagedown"
            ;;
        up|right|down|left|start|select|x|y)
            key="$input_name"
            ;;
        a)
            key="$input_name"
            ;;
        b)
            key="$input_name"
            ;;
        *)
            return
            ;;
    esac
    local es_conf="/home/emubox/.emulationstation/es_input.cfg"

    # add or update element
    if [[ $(xmlstarlet sel -t -v "count(/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"]/input[@name=\"$key\"])" "$es_conf") -eq 0 ]]; then
        xmlstarlet ed -L -s "/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"]" -t elem -n newinput -v "" \
            -i //newinput -t attr -n "name" -v "$key" \
            -i //newinput -t attr -n "type" -v "$input_type" \
            -i //newinput -t attr -n "id" -v "$input_id" \
            -i //newinput -t attr -n "value" -v "$input_value" \
            -r //newinput -v input \
            "$es_conf"
    else  # if device already exists, update it
        xmlstarlet ed -L \
            -u "/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"]/input[@name=\"$key\"]/@type" -v "$input_type" \
            -u "/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"]/input[@name=\"$key\"]/@id" -v "$input_id" \
            -u "/inputList/inputConfig[@deviceGUID=\"$DEVICE_GUID\"]/input[@name=\"$key\"]/@value" -v "$input_value" \
            "$es_conf"
    fi
}

function onstart_emulationstation_keyboard() {
    onstart_emulationstation_joystick "$@"
}

function map_emulationstation_keyboard() {
    map_emulationstation_joystick "$@"
}


data/usr/share/emubox/emulationstation/scripts/configscripts/retroarch.sh
#!/usr/bin/env bash

# This file is part of The RetroPie Project
#
# The RetroPie Project is the legal property of its developers, whose names are
# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
#
# See the LICENSE.md file at the top-level directory of this distribution and
# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
#

function onstart_retroarch_joystick() {
    iniConfig " = " '"' "/home/emubox/.config/retroarch/retroarch.cfg"
    iniGet "input_joypad_driver"
    local input_joypad_driver="$ini_value"
    if [[ -z "$input_joypad_driver" ]]; then
        input_joypad_driver="udev"
    fi

    _retroarch_select_hotkey=1

    iniConfig " = " "\"" "/tmp/tempconfig.cfg"
    iniSet "input_device" "$DEVICE_NAME"
    iniSet "input_driver" "$input_joypad_driver"
}

function onstart_retroarch_keyboard() {
    iniConfig " = " '"' "/home/emubox/.config/retroarch/retroarch.cfg"

    _retroarch_select_hotkey=1

    declare -Ag retroarchkeymap
    # SDL codes from https://wiki.libsdl.org/SDLKeycodeLookup
    retroarchkeymap["1073741904"]="left"
    retroarchkeymap["1073741903"]="right"
    retroarchkeymap["1073741906"]="up"
    retroarchkeymap["1073741905"]="down"
    retroarchkeymap["13"]="enter"
    retroarchkeymap["1073741912"]="kp_enter"
    retroarchkeymap["9"]="tab"
    retroarchkeymap["1073741897"]="insert"
    retroarchkeymap["127"]="del"
    retroarchkeymap["1073741901"]="end"
    retroarchkeymap["1073741898"]="home"
    retroarchkeymap["1073742053"]="rshift"
    retroarchkeymap["1073742049"]="shift"
    retroarchkeymap["1073742048"]="ctrl"
    retroarchkeymap["1073742050"]="alt"
    retroarchkeymap["32"]="space"
    retroarchkeymap["27"]="escape"
    retroarchkeymap["43"]="add"
    retroarchkeymap["45"]="subtract"
    retroarchkeymap["1073741911"]="kp_plus"
    retroarchkeymap["1073741910"]="kp_minus"
    retroarchkeymap["1073741882"]="f1"
    retroarchkeymap["1073741883"]="f2"
    retroarchkeymap["1073741884"]="f3"
    retroarchkeymap["1073741885"]="f4"
    retroarchkeymap["1073741886"]="f5"
    retroarchkeymap["1073741887"]="f6"
    retroarchkeymap["1073741888"]="f7"
    retroarchkeymap["1073741889"]="f8"
    retroarchkeymap["1073741890"]="f9"
    retroarchkeymap["1073741891"]="f10"
    retroarchkeymap["1073741892"]="f11"
    retroarchkeymap["1073741893"]="f12"
    retroarchkeymap["48"]="num0"
    retroarchkeymap["49"]="num1"
    retroarchkeymap["50"]="num2"
    retroarchkeymap["51"]="num3"
    retroarchkeymap["52"]="num4"
    retroarchkeymap["53"]="num5"
    retroarchkeymap["54"]="num6"
    retroarchkeymap["55"]="num7"
    retroarchkeymap["56"]="num8"
    retroarchkeymap["57"]="num9"
    retroarchkeymap["1073741899"]="pageup"
    retroarchkeymap["1073741902"]="pagedown"
    retroarchkeymap["1073741922"]="keypad0"
    retroarchkeymap["1073741913"]="keypad1"
    retroarchkeymap["1073741914"]="keypad2"
    retroarchkeymap["1073741915"]="keypad3"
    retroarchkeymap["1073741916"]="keypad4"
    retroarchkeymap["1073741917"]="keypad5"
    retroarchkeymap["1073741918"]="keypad6"
    retroarchkeymap["1073741919"]="keypad7"
    retroarchkeymap["1073741920"]="keypad8"
    retroarchkeymap["1073741921"]="keypad9"
    retroarchkeymap["46"]="period"
    retroarchkeymap["1073741881"]="capslock"
    retroarchkeymap["1073741907"]="numlock"
    retroarchkeymap["8"]="backspace"
    retroarchkeymap["42"]="multiply"
    retroarchkeymap["47"]="divide"
    retroarchkeymap["1073741894"]="print_screen"
    retroarchkeymap["1073741895"]="scroll_lock"
    retroarchkeymap["96"]="backquote"
    retroarchkeymap["1073741896"]="pause"
    retroarchkeymap["39"]="quote"
    retroarchkeymap["44"]="comma"
    retroarchkeymap["45"]="minus"
    retroarchkeymap["47"]="slash"
    retroarchkeymap["59"]="semicolon"
    retroarchkeymap["61"]="equals"
    retroarchkeymap["91"]="leftbracket"
    retroarchkeymap["92"]="backslash"
    retroarchkeymap["93"]="rightbracket"
    retroarchkeymap["1073741923"]="kp_period"
    retroarchkeymap["1073741927"]="kp_equals"
    retroarchkeymap["1073742052"]="rctrl"
    retroarchkeymap["1073742054"]="ralt"
    retroarchkeymap["97"]="a"
    retroarchkeymap["98"]="b"
    retroarchkeymap["99"]="c"
    retroarchkeymap["100"]="d"
    retroarchkeymap["101"]="e"
    retroarchkeymap["102"]="f"
    retroarchkeymap["103"]="g"
    retroarchkeymap["104"]="h"
    retroarchkeymap["105"]="i"
    retroarchkeymap["106"]="j"
    retroarchkeymap["107"]="k"
    retroarchkeymap["108"]="l"
    retroarchkeymap["109"]="m"
    retroarchkeymap["110"]="n"
    retroarchkeymap["111"]="o"
    retroarchkeymap["112"]="p"
    retroarchkeymap["113"]="q"
    retroarchkeymap["114"]="r"
    retroarchkeymap["115"]="s"
    retroarchkeymap["116"]="t"
    retroarchkeymap["117"]="u"
    retroarchkeymap["118"]="v"
    retroarchkeymap["119"]="w"
    retroarchkeymap["120"]="x"
    retroarchkeymap["121"]="y"
    retroarchkeymap["122"]="z"

    # special case for disabled hotkey
    retroarchkeymap["0"]="nul"
}

function map_retroarch_joystick() {
    local input_name="$1"
    local input_type="$2"
    local input_id="$3"
    local input_value="$4"

    local keys
    case "$input_name" in
        up)
            keys=("input_up")
            ;;
        down)
            keys=("input_down")
            ;;
        left)
            keys=("input_left" "input_state_slot_decrease")
            ;;
        right)
            keys=("input_right" "input_state_slot_increase")
            ;;
        a)
            keys=("input_a")
            ;;
        b)
            keys=("input_b" "input_reset")
            ;;
        x)
            keys=("input_x" "input_menu_toggle")
            ;;
        y)
            keys=("input_y")
            ;;
        leftbottom|leftshoulder)
            keys=("input_l" "input_load_state")
            ;;
        rightbottom|rightshoulder)
            keys=("input_r" "input_save_state")
            ;;
        lefttop|lefttrigger)
            keys=("input_l2")
            ;;
        righttop|righttrigger)
            keys=("input_r2")
            ;;
        leftthumb)
            keys=("input_l3")
            ;;
        rightthumb)
            keys=("input_r3")
            ;;
        start)
            keys=("input_start" "input_exit_emulator")
            ;;
        select)
            keys=("input_select")
            ;;
        leftanalogleft)
            keys=("input_l_x_minus")
            ;;
        leftanalogright)
            keys=("input_l_x_plus")
            ;;
        leftanalogup)
            keys=("input_l_y_minus")
            ;;
        leftanalogdown)
            keys=("input_l_y_plus")
            ;;
        rightanalogleft)
            keys=("input_r_x_minus")
            ;;
        rightanalogright)
            keys=("input_r_x_plus")
            ;;
        rightanalogup)
            keys=("input_r_y_minus")
            ;;
        rightanalogdown)
            keys=("input_r_y_plus")
            ;;
        hotkeyenable)
            keys=("input_enable_hotkey")
            _retroarch_select_hotkey=0
            if [[ "$input_type" == "key" && "$input_id" == "0" ]]; then
                return
            fi
            ;;
        *)
            return
            ;;
    esac

    local key
    local value
    local type
    for key in "${keys[@]}"; do
        case "$input_type" in
            hat)
                type="btn"
                value="h$input_id$input_name"
                ;;
            axis)
                type="axis"
                if [[ "$input_value" == "1" ]]; then
                    value="+$input_id"
                else
                    value="-$input_id"
                fi
                ;;
            *)
                type="btn"
                value="$input_id"
                ;;
        esac
        if [[ "$input_name" == "select" && "$_retroarch_select_hotkey" -eq 1 ]]; then
            _retroarch_select_type="$type"
        fi
        key+="_$type"
        iniSet "$key" "$value"
    done
}

function map_retroarch_keyboard() {
    local input_name="$1"
    local input_type="$2"
    local input_id="$3"
    local input_value="$4"

    local key
    case "$input_name" in
        up)
            keys=("input_player1_up")
            ;;
        down)
            keys=("input_player1_down")
            ;;
        left)
            keys=("input_player1_left" "input_state_slot_decrease")
            ;;
        right)
            keys=("input_player1_right" "input_state_slot_increase")
            ;;
        a)
            keys=("input_player1_a")
            ;;
        b)
            keys=("input_player1_b" "input_reset")
            ;;
        x)
            keys=("input_player1_x" "input_menu_toggle")
            ;;
        y)
            keys=("input_player1_y")
            ;;
        leftbottom|leftshoulder)
            keys=("input_player1_l")
            ;;
        rightbottom|rightshoulder)
            keys=("input_player1_r")
            ;;
        lefttop|lefttrigger)
            keys=("input_player1_l2")
            ;;
        righttop|righttrigger)
            keys=("input_player1_r2")
            ;;
        leftthumb)
            keys=("input_player1_l3")
            ;;
        rightthumb)
            keys=("input_player1_r3")
            ;;
        start)
            keys=("input_player1_start" "input_exit_emulator")
            ;;
        select)
            keys=("input_player1_select")
            ;;
        hotkeyenable)
            keys=("input_enable_hotkey")
            _retroarch_select_hotkey=0
            ;;
        *)
            return
            ;;
    esac

    for key in "${keys[@]}"; do
        iniSet "$key" "${retroarchkeymap[$input_id]}"
    done
}

function onend_retroarch_joystick() {
    # if $_retroarch_select_hotkey is set here, then there was no hotkeyenable button
    # in the configuration, so we should use the select button as hotkey enable if set
    if [[ "$_retroarch_select_hotkey" -eq 1 ]]; then
        iniGet "input_select_${_retroarch_select_type}"
        [[ -n "$ini_value" ]] && iniSet "input_enable_hotkey_${_retroarch_select_type}" "$ini_value"
    fi

    # hotkey sanity check
    # remove hotkeys if there is no hotkey enable button
    if ! grep -q "input_enable_hotkey" /tmp/tempconfig.cfg; then
        local key
        for key in input_state_slot_decrease input_state_slot_increase input_reset input_menu_toggle input_load_state input_save_state input_exit_emulator; do
            sed -i "/$key/d" /tmp/tempconfig.cfg
        done
    fi

    # disable any auto configs for the same device to avoid duplicates
    local file
    local dir="/home/emubox/.config/retroarch/autoconfig"
    while read -r file; do
        mv "$file" "$file.bak"
    done < <(grep -Fl "\"$DEVICE_NAME\"" "$dir/"*.cfg 2>/dev/null)

    # sanitise filename
    file="${DEVICE_NAME//[\?\<\>\\\/:\*\|]/}.cfg"

    if [[ -f "$dir/$file" ]]; then
        mv "$dir/$file" "$dir/$file.bak"
    fi
    mv "/tmp/tempconfig.cfg" "$dir/$file"
}

function onend_retroarch_keyboard() {
    if [[ "$_retroarch_select_hotkey" -eq 1 ]]; then
        iniGet "input_player1_select"
        iniSet "input_enable_hotkey" "$ini_value"
    fi

    # hotkey sanity check
    # remove hotkeys if there is no hotkey enable button
    iniGet "input_enable_hotkey"
    if [[ -z "$ini_value" || "$ini_value" == "nul" ]]; then
        iniSet "input_state_slot_decrease" ""
        iniSet "input_state_slot_increase" ""
        iniSet "input_reset" ""
        iniSet "input_menu_toggle" "f1"
        iniSet "input_load_state" ""
        iniSet "input_save_state" ""
        iniSet "input_exit_emulator" "escape"
        iniSet "input_shader_next" ""
        iniSet "input_shader_prev" ""
        iniSet "input_rewind" ""
    fi
}
